sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala

```diff
@@ -642,81 +643,77 @@ case class GpuSum(child: Expression,
       with GpuAggregateWindowFunction
       with GpuRunningWindowFunction {
 
-  private lazy val cudfSum = AttributeReference("sum", resultType)()
-  private lazy val isEmpty = AttributeReference("isEmpty", BooleanType, nullable = false)()
-  private lazy val isOverflow = AttributeReference("isOverflow", BooleanType, nullable = false)()
   private lazy val zeroDec = {
     val dt = resultType.asInstanceOf[DecimalType]
     GpuLiteral(Decimal(0, dt.precision, dt.scale), dt)
   }
 
-  override lazy val inputProjection: Seq[Expression] = resultType match {
-    case _: DecimalType => Seq(child, GpuIsNull(child))
-    case _ => Seq(child)
+  override lazy val initialValues: Seq[GpuLiteral] = resultType match {
+    case _: DecimalType => Seq(zeroDec, GpuLiteral(true, BooleanType))
+    case _ => Seq(GpuLiteral(null, resultType))
   }
 
   // we need to cast to `resultType` here, since Spark is not widening types
   // as done before Spark 3.2.0. See CudfSum for more info.
-  override lazy val preUpdate: Seq[Expression] = resultType match {
+  override lazy val inputProjection: Seq[Expression] = resultType match {
     case _: DecimalType =>
-      // Spark tracks null columns through isEmpty for decimal.
-      // TODO should we cast the sum to a type with one larger precision first (if possible)
-      //  that way we can get a larger range to detect overflows 10 billion entries instead of
-      //  1 billion
-      Seq(GpuIf(isEmpty, zeroDec, GpuCast(cudfSum, resultType)), isEmpty)
-    case _ => Seq(GpuCast(cudfSum, resultType))
+      // Spark tracks null columns through a second column isEmpty for decimal.
+      Seq(GpuIf(GpuIsNull(child), zeroDec, GpuCast(child, resultType)), GpuIsNull(child))
+    case _ => Seq(GpuCast(child, resultType))
   }
 
-  override lazy val updateExpressions: Seq[Expression] = resultType match {
-    case _: DecimalType => Seq(new CudfSum(cudfSum), new CudfMin(isEmpty))
-    case _ => Seq(new CudfSum(cudfSum))
+  override lazy val updateAggregates: Seq[CudfAggregate] = resultType match {
+    case _: DecimalType => Seq(new CudfSum(resultType), new CudfMin(BooleanType))
+    case _ => Seq(new CudfSum(resultType))
   }
 
+  private lazy val postUpdateSum = updateAggregates.head.attr
```

Comment:
Minor but I think it can be cleaner (than referring to the members of `updateAggregates` positionally) 

Instead of referencing the `attr` here, I'd propose we define:
```
lazy val updateSum = new CudfSum(resultType)
lazy val updateIsEmpty = new CudfMin(BooleanType)
```

Then `updateAggregates` refers to these as such:
```
  override lazy val updateAggregates: Seq[CudfAggregate] = resultType match {
    case _: DecimalType => Seq(updateSum, updateIsEmpty)
    case _ => Seq(updateSum)
  }
```

Allowing us to pick the `attr` member in `postUpdate` like so:

```
  override lazy val postUpdate: Seq[Expression] = resultType match {
    case dt: DecimalType =>
      Seq(GpuCheckOverflow(updateSum.attr, dt, !failOnErrorOverride), updateIsEmpty.attr)
    case _ => updateSum.attr
  }
```

Reply:
This pattern can be applied in other places where the attributes are picked using head or using an index in other `GpuAggregateFunction` impls.